### DISCUSSION TITLE
Can Liquid includes generate different output in different contexts?

### DISCUSSION THREAD

**Discussion by robweychert** [11ty/eleventy#2570] [Q&A] [answered]
This question might be more broadly about Liquid than Eleventy’s specific implementation of it, but:

My site’s posts have a lot of figures that incorporate inline SVG, which works great on the main site, but I want to replace the SVG with PNG when those posts are published in my RSS feed and newsletter. I thought it would be a simple matter of a conditional in the figure include, something like this:

```liquid
{% if layout == 'layouts/rss.liquid' or layout == 'layouts/newsletter.liquid' %}
  {{ PNG content here }}
{% else %}
  {{ SVG content here }}
{% endif %}
```

But in all contexts, it assumes the layout in question is the one specified in the post’s front matter.

I also tried setting variables in each layout like this:

```liquid
{% assign offsite = true %}
```

And then tried this in the figure include:

```liquid
{% if offsite %}
  {{ PNG content here }}
{% else %}
  {{ SVG content here }}
{% endif %}
```

But I guess includes can’t read variables in template files?

Anyway, I’m out of ideas for the moment. Suggestions would be much appreciated!

**Comment by pdehaan** [comment, marked as answer]
Yeah, I think I've found `{% assign %}` will put the variable in the current page's scope, but not in layouts or other files. But setting things in front matter will propagate or data directory files or probably even `eleventyComputed` will work.

Not sure if you have a repo w/ a simple test case where I can take a look at your use case.

**Reply by pdehaan** [reply]
Maybe I'm misunderstanding something. Here's a very weird demo of 5 pages w/ different layouts; some set `layout`s via front matter, some via `eleventyComputed`, some without `layout`s that use fallback directory data files. Some use include files via the shiny new LiquidJS [`{% render %}`](https://liquidjs.com/tags/render.html) tag, others use the deprecated [`{% include %}`](https://liquidjs.com/tags/include.html) tag.

https://github.com/pdehaan/11ty-2570

**Reply by robweychert** [reply]
I’m new to Eleventy and mostly trying to ease into it by pretending it’s Jekyll, including using the (apparently deprecated!) `{% include %}` tag. Thank you for the links and for putting together this demo! I’ll dig through it, see what I can make work, and let you know how it goes.

**Reply by pdehaan** [reply]
@robweychert if you have a public repo, we can take a look at your code and help you get set up.

**Reply by robweychert** [reply]
@pdehaan I appreciate it! The repo is private for now, and this issue is pretty much the last piece of the puzzle for this implementation (which has otherwise gone pretty smoothly), so I’m going to try to tackle it with the info you provided before making the repo public and asking for more help. Thanks again!

**Reply by robweychert** [reply]
@pdehaan I looked through your demo, and it doesn’t quite replicate the problem. If you’d like to take a look, I put together a simplified version of my setup here:

https://github.com/robweychert/11ty-include-test

Here’s more detail on what’s happening:

* Posts are published at their permalinks using the `post.liquid` layout.
* Instances of the `figure.liquid` include are featured in the posts, but not in the layout templates.
* `figure.liquid` specifies using SVG when being parsed by `post.liquid`, and PNG when parsed by any other layout.
* Full post content for all posts is aggregated and published in `index.rss` at the site root, which uses the `rss.liquid` layout.
* But rather than showing the intended PNG, the `index.rss` output shows the SVG. `figure.liquid` can’t seem to see past the front matter of the post it appears in, even when a copy of that post is parsed by `rss.liquid`.

**Reply by pdehaan** [reply]
Ah sorry, I completely missed this notification...
Bad news is, I _think_ that is the intended behavior (although I'm struggling to find somewhere in the docs that explains the `.contents` or `.layoutContent` properties when using collections &mdash; it's a hard string to search for).

But _my_ understanding of it is, the `post.data.content` is a cached value from the template. So you're using THAT file's layout, and you can't override it and cause the entire page to re-render itself in a different context; if that makes sense (and I barely understand myself).

I also tried w/ a shortcode instead of an `{% include %}` but got the same result.
I do have a weird (and possibly expensive) solution, but at least it's something&hellip; Basically I use [cheerio](https://cheerio.js.org/index.html) to parse the `post.data.content` and then replace the SVG with a specified fallback image within the layouts/rss.liquid layout. This could probably be cleaned up, but it's a starting point (and possibly a terrible idea).

```js
// .eleventy.js
const cheerio = require("cheerio");

/**
 * @typedef {import('@11ty/eleventy/src/UserConfig')} EleventyConfig
 * @typedef {ReturnType<import('@11ty/eleventy/src/defaultConfig')>} EleventyReturnValue
 * @type {(eleventyConfig: EleventyConfig) => EleventyReturnValue}
 */
module.exports = function (eleventyConfig) {

  eleventyConfig.addFilter("svg_to_img", function (html) {
    const KEY = "data-rss-img";
    // Return input HTML if there isn't specified KEY marker (to avoid parsing w/ cheerio).
    if (!html.includes(KEY)) {
      return html;
    }
    // https://github.com/cheeriojs/cheerio/issues/1031#issuecomment-748677236
    const isDocument = false;
    const $ = cheerio.load(html, null, isDocument);
    // Find all [data-rss-img] attributes.
    for (const el of $(`[${KEY}]`)) {
      const _el = $(el);
      // Append our new `img` tag after the current element.
      $(`<img src="${ _el.attr(KEY) }" alt="" />`).insertAfter(_el);
      // Remove the previous element.
      _el.remove();
    }
    return $.html();
  });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

### layouts/rss.liquid
```liquid
{% assign posts = collections.post | reverse %}
{% for post in posts %}
  <section data-layout="{{ post.data.layout }}">
  {{ post.data.content | svg_to_img }}
  </section>
{% endfor %}
```

### posts/post1.md
````liquid
---
title: Post 1
permalink: /post1/
---

Best of times

{% render "figure", rss_img: "/post1.png" %}

- one fish
- two fish
- red fish
- blue fish

{% render "figure", rss_img: "/post1b.png" %}

```js
// code block
```
````

### _includes/figure.liquid
Note how we're specifying a `data-rss-img` data attribute with the value of our `{{ rss_img }}` variable via [`{% render %}`](https://liquidjs.com/tags/render.html) or [the deprecated LiquidJS] [`{% include %}`](https://liquidjs.com/tags/include.html) tag.

> _**Deprecated**_
> _[The [`include`](https://liquidjs.com/tags/include.html) tag] is deprecated, use [render](https://liquidjs.com/tags/render.html) tag instead, which contains all the features of include and provides better encapsulation._

```liquid
<svg width="200" height="200" xmlns="http://www.w3.org/2000/svg" data-rss-img="{{ rss_img }}">
  <path d="M10 10"/>
  <!-- Points -->
  <circle cx="10" cy="10" r="2" fill="red"/>
</svg>
```

## OUTPUT

### _site/post1/index.html
```html
<h1>Post 1</h1>

<p>Best of times</p>
<svg
  width="200"
  height="200"
  xmlns="http://www.w3.org/2000/svg"
  data-rss-img="/post1.png"
>
  <path d="M10 10" />
  <!-- Points -->
  <circle cx="10" cy="10" r="2" fill="red" />
</svg>
<ul>
  <li>one fish</li>
  <li>two fish</li>
  <li>red fish</li>
  <li>blue fish</li>
</ul>
<svg
  width="200"
  height="200"
  xmlns="http://www.w3.org/2000/svg"
  data-rss-img="/post1b.png"
>
  <path d="M10 10" />
  <!-- Points -->
  <circle cx="10" cy="10" r="2" fill="red" />
</svg>
<pre><code class="language-js">// code block
</code></pre>
```

### _site/index.rss

```xml
<section data-layout="layouts/post.liquid">
  <h1>Post 2</h1>
  <img src="/post2.gif" />
</section>

<section data-layout="layouts/post.liquid">
  <h1>Post 1</h1>
  <p>Best of times</p>
  <img src="/post1.png" />
  <ul>
    <li>one fish</li>
    <li>two fish</li>
    <li>red fish</li>
    <li>blue fish</li>
  </ul>
  <img src="/post1b.png" />
  <pre><code class="language-js">// code block
</code></pre>
</section>
```

**Reply by pdehaan** [reply]
Like I say, **cheerio** is a pretty big hammer, but #yolo. There might be other ways to do this.

Other idea that came to mind was a custom 11ty [transform](https://www.11ty.dev/docs/config/#transforms) that only modifies RSS files and does something similar re:DOM manipulation.

**Reply by robweychert** [reply]
@pdehaan Thanks for tinkering! We’re getting somewhat into over-my-head territory, and as you said, cheerio seems a bit overpowered for the task, but a custom 11ty transform seems like it might be the way to go. My post SVG output includes all the data the RSS PNG version would need (`src`, `width`, `height`, `alt`, etc), so if a custom transform can manipulate the DOM of the RSS output, that may well do the trick. I’ll give it a whirl and let you know how it goes.

**Reply by pdehaan** [reply]
To be fair, I think the 11ty transform would still need cheerio or something to do DOM manipulation. The transforms get the "final" HTML as a string, so you'd need some way to search the input for things you want to replace (SVGs to images) and then rewrite out the DOM as a string so it can be saved to disk.

Or maybe you can use something like https://github.com/rbren/rss-parser which apparently has a `.parseString()` API. Although you'd still need to update the elements you want and then write back as an .RSS file (not a .JSON file).

But yeah, keep me updated; sounds interesting!

**Reply by robweychert** [reply]
> The transforms get the "final" HTML as a string

I *thought* the DOM manipulation thing sounded too easy. 🙂

**Reply by pdehaan** [reply]
Bonus points if you can post a copy of your sample RSS file somewhere I can peek at it.

Curious if some regular expressions might be an option.

**Reply by robweychert** [reply]
Regex would probably work. I have very little experience with it, but this isn’t the only project I’m working on that could make use of it, so maybe it’s time to finally dig in. Here’s the basic markup of my figures; first the onsite SVG:

```html
<figure id="fig01" class="article__fig">
  <svg width="100" height="100" viewBox="0 0 100 100" aria-labelledby="fig01-desc">
    <desc id="fig01-desc>Here is the alt text.</desc>
    <a class="hide-on-fallback">
      [ SVG data goes here ]
    </a>
    <image src="fig01.png" width="100" height="100 xlink:href=""/>
  </svg>
  <figcaption class="article__caption">
    <strong class="article__fig-label">Fig. 1:</strong> Here is the caption.
  </figcaption>
</figure>
```

And here’s the PNG version that appears in the RSS:

```html
<figure>
  <img src="https://foo.foo/fig01.png" width="100" height="100" alt="Here is the alt text.">
  <figcaption>
    <strong>Fig. 1:</strong> Here is the caption.
  </figcaption>
</figure>
```

I think regex would be able to turn the first into the second?

**Reply by pdehaan** [reply]
RegExp can do _everything_. It's just a matter of gross it is, and how incredibly brittle it is in the real world with real data.
> _Spoiler alert: "It got pretty gross."_

If you were hoping for a solution with a bunch of non-greedy regular expressions w/ [named capture groups](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Backreferences), you're in luck!

So, turns out I needed _three_ regular expressions:
1. Fetch the `/aria-labelledby/`key.
2. Fetch the `<desc>` tag with the matching aria id.
3. Fetch the SVG `<image>` tag and try and extract the `src` attribute and then the rest of the attributes.

Then use `String#replace()` to swap the `<svg>` tag with our custom `<img>` tag.

Overall, it's pretty brittle. I have no idea if it'd work if a post had multiple figures, or if a `<desc>` spanned more than a single line, or if the `<image>` attributes spanned more than a single line (or had attributes in not this exact order). Or if you are using this in RSS and each page uses the same `aria-labelled-by="fig01-desc"` pattern, and nothing is unique

```js
const figHtml = `
<figure id="fig01" class="article__fig">
  <svg width="100" height="100" viewBox="0 0 100 100" aria-labelledby="fig01-desc">
    <desc id="fig01-desc">Here is the alt text.</desc>
    <a class="hide-on-fallback">
      [ SVG data goes here ]
    </a>
    <image src="fig01.png" width="100" height="100" xlink:href=""/>
  </svg>
  <figcaption class="article__caption">
    <strong class="article__fig-label">Fig. 1:</strong> Here is the caption.
  </figcaption>
</figure>
`;

const { altId } = figHtml.match(/aria-labelledby="(?<altId>.*?)"/)?.groups;
const { alt } = figHtml.match(new RegExp(`<desc id="${altId}">(?<alt>.*?)</desc>`))?.groups;
const { src, attrs } = figHtml.match(/<image src="(?<src>.*?)" (?<attrs>.*?)\/>/)?.groups;

const figRss = figHtml.replace(/<svg .*?>(\n|.)*?<\/svg>/m, `<img src="https://foo.bar/baz/${src}" ${attrs} alt="${alt}" />`);

console.log(figRss);
```

### OUTPUT
```html
<figure id="fig01" class="article__fig">
  <img src="https://foo.bar/baz/fig01.png" width="100" height="100" xlink:href="" alt="Here is the alt text." />
  <figcaption class="article__caption">
    <strong class="article__fig-label">Fig. 1:</strong> Here is the caption.
  </figcaption>
</figure>
```

**Reply by robweychert** [reply]
There will indeed be multiple posts in the feed, and multiple figures per post. But this is definitely a start! *[cracks knuckles]*

**Reply by robweychert** [reply]
Success! Here’s the simplified version of where I landed:

### The original `index.rss` content

```html
<figure id="fig01" class="article__fig">
  <svg width="100" height="100" viewBox="0 0 100 100" aria-labelledby="fig01-desc">
    <desc id="fig01-desc">Here is the alt text.</desc>
    <a class="hide-on-fallback">
      [ SVG data goes here ]
    </a>
    <image src="fig01.png" width="100" height="100" xlink:href=""/>
  </svg>
  <figcaption class="article__caption">
    <strong class="article__fig-label">Fig. 1:</strong> Here is the caption.
  </figcaption>
</figure>
```

### The transform in `.eleventy.js`

```js
module.exports = function(eleventyConfig) {
  // Adjust markup to be appropriate for RSS feed
  eleventyConfig.addTransform("rssCleanup", function(content, outputPath) {
    if(outputPath && outputPath.endsWith(".rss")) {
      const stringsToRemove = [
        /\sclass\=\".*?\"/g,
        /\sid\=\".*?\"/g,
        `<span>`,
        `</span>`
      ]
      // Remove all classes, ids, and span tags
      for (i = 0; i < stringsToRemove.length; i++) {
        content = content.replaceAll(stringsToRemove[i], ``);
      }
      // Put all svg elements into an array
      const figSvgs = content.match(/\<svg.*?\<\/svg\>/g);
      // For each svg element
      for (i = 0; i < figSvgs.length; i++) {
        // Get the alt text from the desc element
        const alt = figSvgs[i].match(/\<desc\>.*?\<\/desc\>/g).toString().replace(`<desc>`, ``).replace(`</desc>`, ``);
        // Get the image element and convert it to an img element with an absolute path and alt text
        const img = figSvgs[i].match(/\<image.*?\>/g).toString().replace(`<image src="`, `<img src="https://foo.foo/`).replace(`xlink:href=""/>`, `alt="` + alt + `">`);
        // Replace the svg in the content with the img element
        content = content.replace(figSvgs[i], img);
      }
    }
    return content;
  });
};
```

### The final `index.rss` output

```html
<figure>
  <img src="https://foo.foo/fig01.png" width="100" height="100" alt="Here is the alt text.">
  <figcaption>
    <strong>Fig. 1:</strong> Here is the caption.
  </figcaption>
</figure>
```

Thank you so much for your help, @pdehaan!

**Reply by pdehaan** [reply]
Nice! I did actually work on a transform version of the Cheerio idea last night and almost replied. But then waited until I could find a sample RSS file to test on, and found out it didn't really work since the random, sample RSS file I found used `<![CDATA[ ... ]]>` and apparently cheerio wasn't finding those `<figure>` elements anymore once they're nested in CDATA. Which then would bring me back to my original idea of needing to:
1. convert from RSS to JSON, 
2. loop over each `<item>` tag in the feed,
4. convert each `figure > svg` to `figure > img` in each item's `<description>` and `<content:encoded>` tags,
5. convert from JSON back to RSS.

**Reply by robweychert** [reply]
Ah, yep, I’ve never managed to make an RSS feed without incorporating `CDATA`. Glad I didn’t have to go the RSS to JSON to RSS route.

**Reply by pdehaan** [reply]
Yeah, it's a definite tradeoff.
RSS => JSON => RSS can be "safer", but definitely a lot more hoops to jump through.
RegExp can be super nice in some cases but difficult to manage and fragile. For example, if you have a `<span data-attr="true">` in your code, it might not be detected by your RegExps, but the closing tag would still be removed. Similarly, what if there were SOME `<span>` tags you wanted to keep, but certain ones within a `<figure>` that you wanted to remove. It gets hard to restrict the logic. But it all eventually comes down to exact use cases and what this particular RSS feed looks like.

**Reply by robweychert** [reply]
The markup patterns on this site are pretty predictable, so I feel confident about the regex solution, but I hear you on the edge cases!

**Reply by robweychert** [reply]
Although now that you mention it, in the list of strings to remove, I think I’ll make it just a little bit safer by changing that `'<span>'` to `/\<span.*?\>/g`.
